articles/quickstart/backend/java-spring-security5/interactive.md

```diff
@@ -0,0 +1,150 @@
+// build.gradle
+
+plugins {
+    id 'org.springframework.boot' version '2.5.12'
+    id 'io.spring.dependency-management' version '1.0.9.RELEASE'
+}
+
+dependencies {
+    implementation 'org.springframework.boot:spring-boot-starter-web'
+    implementation 'org.springframework.boot:spring-boot-starter-oauth2-resource-server'
+}
+```
+
+If you are using Maven, add the Spring dependencies to your `pom.xml` file:
+
+```xml
+// pom.xml
+
+<parent>
+    <groupId>org.springframework.boot</groupId>
+    <artifactId>spring-boot-starter-parent</artifactId>
+    <version>2.5.12</version>
+    <relativePath/>
+</parent>
+
+<dependencies>
+    <dependency>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-web</artifactId>
+    </dependency>
+    <dependency>
+        <groupId>org.springframework.boot</groupId>
+        <artifactId>spring-boot-starter-oauth2-resource-server</artifactId>
+    </dependency>
+</dependencies>
+```
+
+## Validate the audience {{{ data-action=code data-code="AudienceValidator.java#13:20" }}}
+
+To validate the JWT, you also need to validate that the JWT is intended for your API by checking the `aud` claim of the JWT. Create a new class named `AudienceValidator` that implements the `OAuth2TokenValidator` interface and override the `validate` method to verify whether the required `audience` is present.
+
+## Configure the resource server {{{ data-action=code data-code="SecurityConfig.java" }}}
+
+To configure the application as a Resource Server and validate the JWTs, create a class that extends `WebSecurityConfigurerAdapter`, add the `@EnableWebSecurity` annotation, and override the `configure` method
+
+### Protect API Endpoints
+
+<%= include('../_includes/_api_endpoints') %>
+
+The example below shows how to secure API methods using the `HttpSecurity` object provided in the `configure()` method of the `SecurityConfig` class. Route matchers are used to restrict access based on the level of authorization required:
+
+::: note
+By default, Spring Security will create a `GrantedAuthority` for each scope in the `scope` claim of the JWT. This is what enables using the `hasAuthority("SCOPE_read:messages")` method to restrict access to a valid JWT that contains the `read:messages` scope.
+
+If your use case requires different claims to make authorization decisions, see the [Spring Security Reference Documentation](https://docs.spring.io/spring-security/site/docs/current/reference/htmlsingle/#oauth2resourceserver-authorization-extraction) to learn how to customize the extracted authorities.
+:::
+
+### Configure JWT Validator
+
+Update the `SecurityConfig` class to configure a `JwtDecoder` bean that uses the `AudienceValidator`. The `auth0.audience` value from `application.yml` will be used to validate the `aud` claim. The `issuer` is verified using the `issuer-uri` value from the `application.yml`
+
+## Create the Domain Object {{{ data-action=code data-code="Message.java#1:11" }}}
```

Comment:
I'm curious if everyone knows what is meant by a Domain Object. Also curious if this is something we should explicitly have in our quickstart? Would it make sense to just return an anonymous class or a String instead of an instance of Message from the controller? Avoids the need for terminology like Domain Object, while also avoids solving the issue mentioned in the short sentence below.

Reply:
We have fixed this and also created a sample which just returns a string instead of a POJO - https://github.com/auth0-samples/auth0-spring-security5-api-sample/pull/16

Reply:
Unfortunately we won't be able to remove this :( because all API based quickstarts use a test script and we need to return a JSON object instead of a string. We cannot edit the test script since all the other QS samples depend on it - https://github.com/search?q=org%3Aauth0-samples++api-quickstarts-tests&type=code